File: src/core/units.ts
import os from 'node:os'
import path from 'node:path'

export function formatBytes(bytes: number): string {
  if (!Number.isFinite(bytes) || bytes <= 0) return ''
  const tiers = ['B', 'KB', 'MB', 'GB'] as const
  let value = bytes
  let tier = 0
  while (value >= 1024 && tier < tiers.length - 1) {
    value /= 1024
    tier++
  }
  const formatted =
    value >= 10 || tier === 0 || Number.isInteger(value) ? Math.round(value) : value.toFixed(1)
  return `${formatted} ${tiers[tier]}`
}

export function formatDuration(totalSeconds: number): string {
  if (!Number.isFinite(totalSeconds) || totalSeconds <= 0) return ''
  const s = Math.round(totalSeconds)
  const hrs = Math.floor(s / 3600)
  const mins = Math.floor((s % 3600) / 60)
  const secs = s % 60
  const mm = hrs > 0 ? String(mins).padStart(2, '0') : String(mins)
  const ss = String(secs).padStart(2, '0')
  return hrs > 0 ? `${hrs}:${mm}:${ss}` : `${mm}:${ss}`
}

export function formatSpeed(bytesPerSecond: number): string {
  if (!Number.isFinite(bytesPerSecond) || bytesPerSecond <= 0) return ''
  return `${formatBytes(bytesPerSecond)}/s`
}

export function formatEta(seconds: number): string {
  if (!Number.isFinite(seconds) || seconds <= 0) return ''
  return formatDuration(seconds)
}

// POSIX terminal emulators allocate 2 monospace grid columns for CJK ideographs,
// full-width punctuation, and kana. String.length counts UTF-16 code units (1),
// so naive truncation causes yoga/ink to miscalculate boundaries and wrap unexpectedly.
export function strWidth(str: string): number {
  let width = 0
  for (const ch of str) {
    const code = ch.codePointAt(0) ?? 0
    if (
      (code >= 0x1100 && code <= 0x115f) ||
      (code >= 0x2e80 && code <= 0xa4cf) ||
      (code >= 0xac00 && code <= 0xd7a3) ||
      (code >= 0xf900 && code <= 0xfaff) ||
      (code >= 0xfe10 && code <= 0xfe19) ||
      (code >= 0xfe30 && code <= 0xfe6f) ||
      (code >= 0xff00 && code <= 0xff60) ||
      (code >= 0xffe0 && code <= 0xffe6) ||
      (code >= 0x20000 && code <= 0x2fffd) ||
      (code >= 0x30000 && code <= 0x3fffd)
    ) {
      width += 2
    } else {
      width += 1
    }
  }
  return width
}

export function truncate(text: string, max: number): string {
  if (strWidth(text) <= max) return text
  let res = ''
  let curWidth = 0
  const target = Math.max(1, max - 1)
  for (const ch of text) {
    const w = strWidth(ch)
    if (curWidth + w > target) break
    res += ch
    curWidth += w
  }
  return `${res}…`
}

export function shortenPath(filepath: string, homedir: string, max = 60): string {
  const normFile = path.normalize(filepath)
  const normHome = path.normalize(homedir)
  const isWindows = process.platform === 'win32'
  const isUnderHome = isWindows
    ? normFile.toLowerCase().startsWith(normHome.toLowerCase())
    : normFile.startsWith(normHome)

  const pretty = isUnderHome
    ? `~${normFile.slice(normHome.length).replace(/\\/g, '/')}`
    : normFile
  if (pretty.length <= max) return pretty
  const ext = /\.\w{1,5}$/.exec(pretty)?.[0] ?? ''
  return `${pretty.slice(0, max - ext.length - 1)}…${ext}`
}

export function resolveUserPath(raw: string, homedir = os.homedir()): string {
  let cleaned = raw.trim()
  if (cleaned.startsWith('~')) {
    cleaned = path.join(homedir, cleaned.slice(1).replace(/^[/\\]+/, ''))
  }
  return path.resolve(cleaned)
}

// Word-wrap text into left-flush lines of at most `width` columns.
// Uses strWidth so lines containing East Asian characters don't overflow the panel.
export function wrapText(text: string, width: number): string[] {
  const lines: string[] = []
  let current = ''
  for (const word of text.split(/\s+/).filter(Boolean)) {
    if (!current) {
      current = word
    } else if (strWidth(current) + 1 + strWidth(word) <= width) {
      current += ` ${word}`
    } else {
      lines.push(current)
      current = word
    }
  }
  if (current) lines.push(current)
  return lines
}

File: internal/units/units.go
package units

import (
	"fmt"
	"math"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strings"

	"github.com/mattn/go-runewidth"
)

func FormatBytes(bytes float64) string {
	if math.IsNaN(bytes) || math.IsInf(bytes, 0) || bytes <= 0 {
		return ""
	}
	tiers := []string{"B", "KB", "MB", "GB"}
	val := bytes
	tier := 0
	for val >= 1024 && tier < len(tiers)-1 {
		val /= 1024
		tier++
	}
	var formatted string
	if val >= 10 || tier == 0 || math.Floor(val) == val {
		formatted = fmt.Sprintf("%.0f", math.Round(val))
	} else {
		formatted = fmt.Sprintf("%.1f", val)
	}
	return fmt.Sprintf("%s %s", formatted, tiers[tier])
}

func FormatDuration(totalSeconds float64) string {
	if math.IsNaN(totalSeconds) || math.IsInf(totalSeconds, 0) || totalSeconds <= 0 {
		return ""
	}
	s := int64(math.Round(totalSeconds))
	hrs := s / 3600
	mins := (s % 3600) / 60
	secs := s % 60
	if hrs > 0 {
		return fmt.Sprintf("%d:%02d:%02d", hrs, mins, secs)
	}
	return fmt.Sprintf("%d:%02d", mins, secs)
}

func FormatSpeed(bytesPerSecond float64) string {
	if math.IsNaN(bytesPerSecond) || math.IsInf(bytesPerSecond, 0) || bytesPerSecond <= 0 {
		return ""
	}
	return fmt.Sprintf("%s/s", FormatBytes(bytesPerSecond))
}

func FormatEta(seconds float64) string {
	if math.IsNaN(seconds) || math.IsInf(seconds, 0) || seconds <= 0 {
		return ""
	}
	return FormatDuration(seconds)
}

func Truncate(text string, max int) string {
	if runewidth.StringWidth(text) <= max {
		return text
	}
	target := max - 1
	if target < 1 {
		target = 1
	}
	var b strings.Builder
	curWidth := 0
	for _, r := range text {
		w := runewidth.RuneWidth(r)
		if curWidth+w > target {
			break
		}
		b.WriteRune(r)
		curWidth += w
	}
	b.WriteString("…")
	return b.String()
}

func ShortenPath(filePath string, homeDir string, max ...int) string {
	maxLen := 60
	if len(max) > 0 && max[0] > 0 {
		maxLen = max[0]
	}
	normFile := filepath.Clean(filePath)
	normHome := filepath.Clean(homeDir)

	isUnderHome := false
	if runtime.GOOS == "windows" {
		lowerFile := strings.ToLower(normFile)
		lowerHome := strings.ToLower(normHome)
		isUnderHome = lowerFile == lowerHome ||
			strings.HasPrefix(lowerFile, lowerHome+string(filepath.Separator)) ||
			strings.HasPrefix(lowerFile, lowerHome+"/")
	} else {
		isUnderHome = normFile == normHome || strings.HasPrefix(normFile, normHome+string(filepath.Separator))
	}

	pretty := normFile
	if isUnderHome {
		pretty = "~" + normFile[len(normHome):]
	}

	if len(pretty) <= maxLen {
		return pretty
	}
	ext := filepath.Ext(pretty)
	trimLen := maxLen - len(ext) - 1
	if trimLen < 1 {
		trimLen = 1
	}
	return fmt.Sprintf("%s…%s", pretty[:trimLen], ext)
}

func ResolveUserPath(raw string, customHome ...string) string {
	home := ""
	if len(customHome) > 0 && customHome[0] != "" {
		home = customHome[0]
	} else {
		home, _ = os.UserHomeDir()
	}
	cleaned := strings.TrimSpace(raw)
	if strings.HasPrefix(cleaned, "~") {
		rest := strings.TrimLeft(cleaned[1:], "/\\")
		cleaned = filepath.Join(home, rest)
	}
	abs, err := filepath.Abs(cleaned)
	if err != nil {
		return cleaned
	}
	return abs
}

var whitespaceRegex = regexp.MustCompile(`\s+`)

func WrapText(text string, width int) []string {
	var lines []string
	words := whitespaceRegex.Split(strings.TrimSpace(text), -1)
	current := ""
	for _, word := range words {
		if word == "" {
			continue
		}
		if current == "" {
			current = word
		} else if runewidth.StringWidth(current)+1+runewidth.StringWidth(word) <= width {
			current += " " + word
		} else {
			lines = append(lines, current)
			current = word
		}
	}
	if current != "" {
		lines = append(lines, current)
	}
	return lines
}
